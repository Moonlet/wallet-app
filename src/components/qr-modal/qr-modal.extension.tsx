export const QrModalReaderComponent = (props: any) => null;

export const QrModalReader = QrModalReaderComponent;
