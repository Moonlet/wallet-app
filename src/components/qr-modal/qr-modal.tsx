import React from 'react';
import { View, Modal } from 'react-native';
import { CameraKitCameraScreen, CameraKitCamera } from 'react-native-camera-kit';

export interface IProps {
    onQrCodeScanned: (qrCode: string) => any;
}

interface IState {
    isVisible: boolean;
}

export class QrModalReaderComponent extends React.Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);

        this.state = {
            isVisible: false
        };
    }

    public open = async () => {
        // TODO: hadle exceptions (promise rejection)
        const success = await CameraKitCamera.checkDeviceCameraAuthorizationStatus();
        if (success) {
            this.setState({ isVisible: true });
        }
    };

    public render() {
        return (
            <Modal
                animationType="slide"
                transparent={false}
                visible={this.state.isVisible}
                presentationStyle={'overFullScreen'}
            >
                <View
                    style={{
                        flex: 1,
                        backgroundColor: 'rgba(100,100,100,0.9)'
                    }}
                >
                    <CameraKitCameraScreen
                        testID="CameraKitCameraScreen"
                        actions={{ leftButtonText: 'Cancel' }}
                        onBottomButtonPressed={event =>
                            this.setState({
                                isVisible: false
                            })
                        }
                        scanBarcode={true}
                        onReadCode={event => {
                            this.setState({
                                isVisible: false
                                //    address: event.nativeEvent.codeStringValue
                            });
                            this.props.onQrCodeScanned(event.nativeEvent.codeStringValue);
                        }}
                        showFrame={true}
                        offsetForScannerFrame={10}
                        heightForScannerFrame={500}
                        colorForScannerFrame={'red'}
                    />
                </View>
            </Modal>
        );
    }
}

export const QrModalReader = QrModalReaderComponent;
