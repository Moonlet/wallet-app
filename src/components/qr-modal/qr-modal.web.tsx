import React from 'react';
import { View } from 'react-native';

export class QrModalReaderComponent extends React.Component {
    public render() {
        return <View />;
    }
}

export const QrModalReader = QrModalReaderComponent;
