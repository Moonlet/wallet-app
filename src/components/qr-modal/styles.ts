import { StyleSheet } from 'react-native';
import { ITheme } from '../../core/theme/itheme';

export default (theme: ITheme) =>
    StyleSheet.create({
        container: {
            paddingTop: 0,
            paddingBottom: 18,
            paddingLeft: 0,
            paddingRight: 0,
            backgroundColor: theme.colors.appBackground,
            ...StyleSheet.absoluteFillObject
        }
    });
