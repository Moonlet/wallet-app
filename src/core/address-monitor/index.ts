import { Notifications } from '../messaging/notifications/notifications';
import { Blockchain } from '../blockchain/types';
import DeviceInfo from 'react-native-device-info';
import { IWalletsState, IAccountState } from '../../redux/wallets/state';
import { CONFIG } from '../../config';

const url = CONFIG.dataApiUrl + '/notifications/register';
const monitoredBlockchains = [Blockchain.ETHEREUM, Blockchain.ZILLIQA];

// const getDeviceToken = (): Promise<string> =>
//     Platform.OS === 'android' ? Notifications.getToken() : getApnsToken();

export const updateAddressMonitorTokens = (wallets: IWalletsState) => {
    const addresses = {};

    Object.keys(wallets || []).forEach(id => {
        wallets[id].accounts.forEach((account: IAccountState) => {
            if (monitoredBlockchains.indexOf(account.blockchain) !== -1) {
                if (!addresses[account.blockchain]) {
                    addresses[account.blockchain] = [];
                }

                addresses[account.blockchain].push(account.address);
            }
        });
    });

    const requestAddresses = Object.keys(addresses).map(blockchain => ({
        blockchain,
        addresses: addresses[blockchain]
    }));

    const deviceId = DeviceInfo.getUniqueId();
    Notifications.getToken().then(token => {
        const request = {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                deviceId,
                tokenType: 'fcm',
                token,
                addresses: requestAddresses
            })
        };

        fetch(url, request);
    });
};
