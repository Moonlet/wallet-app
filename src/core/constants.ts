import { Blockchain } from './blockchain/types';

const BLOCKCHAIN_COINS = {
    [Blockchain.ETHEREUM]: 'ETH',
    [Blockchain.ZILLIQA]: 'ZIL',
    [Blockchain.COSMOS]: 'ATOM',
    [Blockchain.STELLAR]: 'XLM'
};

export { BLOCKCHAIN_COINS };
