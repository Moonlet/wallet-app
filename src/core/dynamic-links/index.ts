import dynamicLinks from '@react-native-firebase/dynamic-links';
import { URL, URLSearchParams } from 'react-native-url-polyfill';
import { captureException as SentryCaptureException } from '@sentry/react-native';
import { IReduxState } from '../../redux/state';
import { NavigationService } from '../../navigation/navigation-service';
import { getChainId } from '../../redux/preferences/selectors';
import { setSelectedBlockchain } from '../../redux/wallets/actions';
import { Blockchain } from '../blockchain/types';
import { getNrPendingTransactions, getSelectedAccount } from '../../redux/wallets/selectors';
import { ITokenState } from '../../redux/wallets/state';
import { store } from '../../redux/config';

export class DynamicLinksService {
    private onLinkListener = null;
    private state: IReduxState = null;

    private handlers = {
        '/:blockchain/:token/validator/:validatorAddress': params => {
            this.navigateValidatorAddress(params);
        },
        '/:blockchain/:token/stake/:validatorAddress': params => {
            this.stakeValidatorAddress(params);
        },
        '/:blockchain/stake': params => {
            this.blockchainStake(params);
        },
        '/stake': params => {
            // TODO: implement later on
            // console.log('stakeDefault', params);
        }
    };

    private navigateValidatorAddress(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;
        const validatorAddress = params?.validatorAddress;
        const tokenSymbol = String(params?.token)?.toUpperCase();

        if (blockchain && validatorAddress && tokenSymbol) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            const chainId = getChainId(this.state, blockchain);

            const selectedAccount = getSelectedAccount(this.state);

            if (!selectedAccount) {
                SentryCaptureException(
                    new Error(
                        `Dynamic Links - no selected account, no wallet, blockchain: ${blockchain}, tokenSymbol: ${tokenSymbol}, validatorAddress: ${validatorAddress}`
                    )
                );
                return;
            }

            const token: ITokenState = selectedAccount.tokens[chainId][tokenSymbol];

            NavigationService.navigate('Validator', {
                blockchain,
                undefined, // validator
                accountIndex: selectedAccount.index,
                token,
                canPerformAction: !getNrPendingTransactions(this.state), // hasPendingTransactions
                options: {
                    validatorAddress,
                    tokenSymbol
                }
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - navigateValidatorAddress, invalid params: ${params}`)
            );
        }
    }

    private stakeValidatorAddress(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;
        const validatorAddress = params?.validatorAddress;

        if (blockchain && validatorAddress) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            NavigationService.navigate('SmartScreen', {
                context: {
                    screen: 'StakeNow',
                    step: 'SelectStakeAccount',
                    key: 'select-stake-account',
                    params: {
                        validatorId: validatorAddress
                    }
                },
                navigationOptions: {
                    title: 'Stake now'
                },
                newFlow: true
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - stakeValidatorAddress, invalid params: ${params}`)
            );
        }
    }

    private blockchainStake(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;

        if (blockchain) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            NavigationService.navigate('SmartScreen', {
                context: {
                    screen: 'StakeNow',
                    step: 'StakeSelectValidator',
                    key: 'stake-now-select-validator'
                },
                navigationOptions: {
                    title: 'Stake now'
                },
                newFlow: true
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - blockchainStake, invalid params: ${params}`)
            );
        }
    }

    public async configure(state: IReduxState) {
        this.state = state;

        const dl = dynamicLinks();

        this.onLinkListener = dl.onLink(link => {
            this.handleDynamicLink(link);
        });

        // application is in a background state or has fully quit
        dl.getInitialLink().then(link => {
            this.handleDynamicLink(link);
        });
    }

    // app is in the foreground state
    public async handleDynamicLink(link: any) {
        if (link?.url) {
            this.processUrl(link.url, this.handlers, {
                protocol: ['https:'],
                host: [
                    'moonlet.io',
                    'moonlet.xyz',
                    'moonletwallet.com',
                    'moonlet.app',
                    'moonlet.link'
                ]
            });
        } else {
            // Invalid link url, link:${link}`
        }
    }

    public removeListeners() {
        typeof this.onLinkListener === 'function' && this.onLinkListener();
    }

    private pathPatternToRegexp(pathPattern: string) {
        let exp = '^';

        const pathParts = pathPattern.split('/').filter(Boolean);
        const capturingGroups = [];
        for (const part of pathParts) {
            exp += '\\/';
            if (part.indexOf(':') === 0) {
                exp += '([^\\/]*)';
                capturingGroups.push(part.replace(/^:/, ''));
            } else {
                exp += part;
            }
        }

        exp += '$';
        return {
            regexp: new RegExp(exp),
            capturingGroups
        };
    }

    private parsePath(path: string, pathPatternString: string) {
        const pathPattern = this.pathPatternToRegexp(pathPatternString);
        const match = path.match(pathPattern.regexp);

        if (match) {
            const params = {};
            for (let i = 0; i < pathPattern.capturingGroups.length; i++) {
                params[pathPattern.capturingGroups[i]] = match[i + 1];
            }
            return params;
        }

        return null;
    }

    private processPath(path: string, handlers: any, urlParams = {}) {
        const pathsPatterns = Object.keys(handlers);
        // console.log(path, handlers, urlParams);
        for (const pathPattern of pathsPatterns) {
            if (typeof handlers[pathPattern] === 'function') {
                const params = this.parsePath(path, pathPattern);
                if (params) {
                    return handlers[pathPattern]({
                        ...urlParams,
                        ...params
                    });
                }
            }
        }
    }

    private paramsToObject(entries: any) {
        const result = {};
        for (const [key, value] of entries) {
            // each 'entry' is a [key, value] tupple
            result[key] = value;
        }
        return result;
    }

    private processUrl(url: string, pathHandlers: any = this.handlers, whitelist = {}) {
        const u = new URL(url);

        for (const key of Object.keys(whitelist)) {
            if (u && typeof u[key] === 'string' && whitelist[key].indexOf(u[key]) < 0) {
                SentryCaptureException(
                    new Error(
                        JSON.stringify(
                            `whitelist not matched ${key}=${u[key]} not in whitelist: ${
                                typeof whitelist[key] === 'string'
                                    ? whitelist[key]
                                    : whitelist[key].join(', ')
                            }`
                        )
                    )
                );
                return null;
            }
        }

        let path = u?.pathname;

        if (path && String(path).startsWith('/links')) path = path.replace('/links', '');

        const searchParams = new URLSearchParams(url);

        return this.processPath(path, pathHandlers, this.paramsToObject(searchParams.entries()));
    }
}

export const DynamicLinks = new DynamicLinksService();
