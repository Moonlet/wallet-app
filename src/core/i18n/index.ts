export * from './translation/translate-component';
export * from './translation/translate';
