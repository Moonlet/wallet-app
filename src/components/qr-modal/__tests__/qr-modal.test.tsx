import React from 'react';
import { QrModalReaderComponent, IProps } from '../qr-modal';
import { loadTranslations } from '../../../core/i18n';

import { shallow } from 'enzyme';

const props: IProps = {
    onQrCodeScanned: jest.fn()
};

jest.mock('react-native-camera-kit');

describe('qr code modal ', () => {
    beforeAll(async () => {
        await loadTranslations('en');
    });
    test('renders correctly', () => {
        const wrapper = shallow(<QrModalReaderComponent {...props} />);
        expect(wrapper.debug()).toMatchSnapshot();
    });
});
