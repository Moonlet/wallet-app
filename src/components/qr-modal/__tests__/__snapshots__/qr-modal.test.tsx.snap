// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`qr code modal  renders correctly 1`] = `
"<Component animationType=\\"slide\\" transparent={false} visible={false} presentationStyle=\\"overFullScreen\\" hardwareAccelerated={false}>
  <View style={{...}}>
    <CameraKitCameraScreen testID=\\"CameraKitCameraScreen\\" actions={{...}} onBottomButtonPressed={[Function: onBottomButtonPressed]} scanBarcode={true} onReadCode={[Function: onReadCode]} />
  </View>
</Component>"
`;

exports[`qr code modal  renders correctly 2`] = `
Object {
  "isVisible": false,
}
`;

exports[`qr code modal  renders correctly 3`] = `
Object {
  "isVisible": false,
}
`;

exports[`qr code modal  renders correctly 4`] = `
Object {
  "isVisible": true,
}
`;

exports[`qr code modal  renders correctly 5`] = `
Object {
  "isVisible": false,
}
`;

exports[`qr code modal  renders correctly 6`] = `
Object {
  "isVisible": false,
}
`;
